Add explanation and examples for access rules to wiki

Merseleo opened the pull request:
Closes #331 

## What I have made

<!-- write down what changes have been made, what was removed/added/changed -->
<!-- e.g. I added a service, which does x -->

- I fixed some spelling mistakes in the backend wiki
- I updated outdated examples or descriptions on the testing and contribution sections
- I added an explanation of the access rules used for authorization checks and supported this explanation with examples

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does
not apply.

- [x] I have updated the documentation accordingly and commented my code
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- [x] (for reviewer) I have checked the implementation against the requirements

Commit:
docs: update testing and contributing guidelines and fix spelling mistakes

Commit:
docs: expand contributing guidelines with the access rule system for authorization checks

Slartibartfass2 commented on wiki/Contributing.md:
````diff
@@ -149,14 +149,88 @@ class MainService(
+    .orElseThrow(EntityNotDeletableException())
+```
+
+Custom access rules should be defined in the `service/accessrules/` directory.
````
Add reference
```suggestion
Custom access rules should be defined in the
[`service/accessrules/`](https://github.com/SE-UUlm/snowballr-backend/tree/develop/src/main/kotlin/se/uulm/snowballr/backend/service/accessrules)
directory.
```

Slartibartfass2 commented on wiki/Contributing.md:
````diff
@@ -149,14 +149,88 @@ class MainService(
+
+```kotlin
+// Checks if the current user is the same as the target user and is active.
+fun isSameUserAndActive() = AccessRule<User> { currentUser, entity ->
````
Add the annotation in the example as well.

Merseleo replied:
I added it.

Slartibartfass2 commented on wiki/Contributing.md:
````diff
@@ -149,14 +149,88 @@ class MainService(
+// ...
+```
+
+Combining multiple access rules now allows for more complex permission logic, such as controlling whether an entity can
````
consistency
```suggestion
Combining multiple access rules now allows for more complex authorization logic, such as controlling whether an entity can
```

Slartibartfass2 commented on wiki/Contributing.md:
```diff
@@ -149,14 +149,88 @@ class MainService(
+        .andAlso(isServerAdmin().forTarget())
+        .orElseThrow(EntityNotDeletableException())
+}
+// In EntityService.kt
```
```suggestion

// In EntityService.kt
```

Slartibartfass2 commented on wiki/Contributing.md:
```diff
@@ -149,14 +149,88 @@ class MainService(
+override suspend fun deleteEntity(request: Base.Id): Base.Nothing =
+    withUser(userRepo) { currentUser ->
+        val entityId = parseUUID(request.id, EntityType.ENTITY)
+        val entity = repo.getProjectById(projectId).getOrThrow()
```
```suggestion
        val entity = repo.getEntityById(entityId).getOrThrow()
```

Merseleo replied:
Thank you, good catch.

Slartibartfass2 commented on wiki/Contributing.md:
```diff
@@ -149,14 +149,88 @@ class MainService(
+        // Check authorization using the composed rule
+        isAllowedToDeleteEntity()
+            .forProperty(Entity::id)
+            .orElseThrow { _, entity -> EntityNotDeletableException(entity.description) }
```
Why is this step here if there's already a `orElseThrow` in `isAllowedToDeleteEntity`?

Merseleo replied:
Only for demonstration purpose. This isn't necessary logic, but I want to demonstrate the usage of the `forProperty` helper (i.e. I define my access rule for a UUID and then check for an entity object because I need the details for my exception).

Slartibartfass2 replied:
As discussed, I found this a bit confusing and think that another reader would be confused as well.

Slartibartfass2 commented on wiki/Testing.md:
```diff
@@ -48,11 +48,13 @@ When implementing the layers as described in
+If the _JUnit_ assertions are not easily usable, then use the
+[AssertJ](https://github.com/assertj/assertj?tab=readme-ov-file) test assertions.
 
 All test classes must have the same relative path under `./src/test` as the implementation class under
-`./src/test` and the same name with an additional `Test` at the end. This ensures a clear structure and an easy way to
+`./src/main` and the same name with an additional `Test` at the end. This ensures a clear structure and an easy way to
```
This was correct; the packages in test should mirror main.

Slartibartfass2 commented on wiki/Testing.md:
```diff
@@ -91,6 +93,9 @@ To keep a clean structure, we group all tests in inner classes according to the
 For parameterized tests using gRPC enums, we provide the custom `GrpcEnumSourceTest` annotation, which skips the
 `UNRECOGNIZED` value because using this value to create an object will lead to an exception being thrown.
 
+If you need a test user, set the second argument of the `RepositoryTest` superclass constructor to true.
```
The position might change; using the name should be more robust.
```suggestion
If you need a test user, set the `needsTestUser` argument of the `RepositoryTest` superclass constructor to true.
```

Slartibartfass2 commented on wiki/Testing.md:
```diff
@@ -140,12 +140,11 @@ class CreateExampleTest : MainServiceTest() {
+that the repository always returns a specific object or throws an exception. This way, we can ensure that every call
+made in the service method is also mocked and then test the behavior of the service method according to the behavior of
+our dependencies. For more complex mocks such as how often a method is called, refer to the rich documentation of the
+used mocking library [MockK](https://mockk.io/).
```
```suggestion
that the repository returns a specific object or throws an exception. We then test the behavior of the service method according to the behavior of
our dependencies. For more complex mocks such as how often a method is called, refer to the rich documentation of the
used mocking library [MockK](https://mockk.io/).
```

Slartibartfass2 reviewed (changes requested):
Great work. Some suggestions for improvement.

There's an issue in Getting Started, see the lint checker warning at the bottom. The link shouldn't contain a dot.

Merseleo committed:
docs: improve backend wiki

This includes a better formatting, adding links where advantageous, correcting examples and improve the wording.

Co-authored-by: Felix Schlegel <40503329+Slartibartfass2@users.noreply.github.com>

Merseleo reviewed (commented):
Thank you for the quick review and as @mowi12 and @domienderle also read this wiki section and approved it, this section can be integrated imo.

Commit:
docs: improve example for access rules

Merseleo reviewed (commented):
What do you think of the new example?

Slartibartfass2 reviewed (approved):
Looks great
